docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+
+### Cluster's node failure
+
+This case is when the `radosgw` process stops due to a cluster's node failure.
+The *One Active pipeline* model fully restores the service by
+rescheduling a new POD somewhere in the cluster.
+
+### radosgw's failure due to a bug not related to a certain input pattern
+
+This case is when the `radosgw` process crashes due to a bug not directly
+related to any input type.
+
+Examples:
+
+- Memory leaks
+- Memory corruptions (stack or heap corruptions)
+- Periodic operations or routines not related to an input (GC, measures,
+  telemetry, etc)
+
+For a bug like this, the *One Active pipeline* model could guarantee
+the user's operations *until the next occurrence* of the same malfunctioning.
+
+A definitive solution would be available only when a patch for the issue
+has released.
+
+### radosgw's failure due to a bug related to a certain input pattern
+
+This case is when the `radosgw` process crashes due to a bug directly
+related to a certain input type.
+
+Examples:
+
+- Putting Buckets with some name's pattern
+- Putting Objects that have a certain size
+- Performing an admin operation over a suspended user
+
+For a bug like this, the *One Active pipeline* model could guarantee
+the user's operations under the condition that the crash-triggering
+input is recognized by the user and thus its submission blocked.
+
+### PV Data corruption at application level due to radosgw's bug
+
+This case is when the state on the `PV` corrupts due to a `radosgw`
+bug.
+
+In this unfortunate scenario, the *One Active pipeline* can hardly help.
+A restart could evenly fix the problem or trigger an endless restarting loop.
+
+The fix for this could contemplate an human intervention.
+A definitive solution would be available only when a patch for the issue
+is available.
+
+### PV Data corruption at application level due to radosgw's anomalous exit
+
+This case is when the state on the PV corrupts due to a `radosgw`
+anomalous exit, eg: after a node failure.
+
+This scenario could even be NOT possible because of the safety features
+implemented on the radosgw SFS backend.
+
+The fix for this could contemplate an human intervention.
+
+## Measuring s3gw failures on Kubernetes
+
+After reviewing the cases, we can say that what can be actually solved with
+an HA model with s3gw is when the failure is not dependent to applicative bugs.
+We can handle temporary issues involving the infrastructure that is hosting
+the `radosgw` process.
+
+So, we are mostly interested in observing and measuring the timing of
+Kubernetes when a `radosgw`'s POD is being rescheduled.
+
+Hence, The idea to collect measures regarding a series of restarts
+artificially triggered on the `radosgw`'s POD.
+Obtaining such measures would allow to compute some arbitrary statistics
+for the time required by Kubernetes to restart the `radosgw`'s POD.
+
+## The s3gw Probe
+
+The `s3gw Probe` is a program developed with the purpose of collecting restart
```

Comment:
This strikes me as debug-/test-level operations that might fit in well with the status/admin extensions that Marcel is adding to the s3gw process for the metrics/status/monitoring page?

And the logs should be available from k8s.

Do we need a separate pod for this or can we fold it into our existing processes?

(Testing this ultimately end-to-end will require integration with the LH test suite, if they have automated fault injection testing and recovery timing; that may be far-fetched for the evaluation of this eventual ADR, but the design the ADR should propose to continuously test/validate our HA concept should tie into that framework.)

Reply:
This so called s3gw-probe, has been developed to ease the collecting of stop/start events for this specific research (and potentially to compute some statistics when and if needed) ; it was not meant (at least it wasn't my initial though) to make an official tool with a more authoritative role outside the scope of this research. 
If we want to make all the tests and measures we will do for this research officially replicable, I agree that we must devote a coordination effort to understand where to put all the probes and what kind of sources do we want to use to collect the events (e.g. the logs).

Reply:
Right, ultimately I think we want to somehow integrate this with the longhorn-tests, I suspect, if that's feasible at all, so it can truly be tested end-to-end at the system level.